osutil/sync_dir_test.go

```diff
@@ -0,0 +1,175 @@
+	c.Assert(changed, HasLen, 0)
+	c.Assert(removed, HasLen, 0)
+	// The content is correct
+	content, err := ioutil.ReadFile(path.Join(s.dir, "expected.snap"))
+	c.Assert(err, IsNil)
+	c.Assert(content, DeepEquals, []byte(`expected`))
+	// The permissions are correct
+	stat, err := os.Stat(name)
+	c.Assert(err, IsNil)
+	c.Assert(stat.Mode().Perm(), Equals, os.FileMode(0600))
+}
+
+func (s *EnsureDirStateSuite) TestCreatesMissingFiles(c *C) {
+	name := filepath.Join(s.dir, "missing.snap")
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{
+		"missing.snap": {Content: []byte(`content`), Mode: 0600},
+	})
+	c.Assert(err, IsNil)
+	// Created file is reported
+	c.Assert(changed, DeepEquals, []string{"missing.snap"})
+	c.Assert(removed, HasLen, 0)
+	// The content is correct
+	content, err := ioutil.ReadFile(name)
+	c.Assert(err, IsNil)
+	c.Assert(content, DeepEquals, []byte(`content`))
+	// The permissions are correct
+	stat, err := os.Stat(name)
+	c.Assert(err, IsNil)
+	c.Assert(stat.Mode().Perm(), Equals, os.FileMode(0600))
+}
+
+func (s *EnsureDirStateSuite) TestRemovesUnexpectedFiless(c *C) {
+	name := filepath.Join(s.dir, "evil.snap")
+	err := ioutil.WriteFile(name, []byte(`evil text`), 0600)
+	c.Assert(err, IsNil)
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{})
+	c.Assert(err, IsNil)
+	// Removed file is reported
+	c.Assert(changed, HasLen, 0)
+	c.Assert(removed, DeepEquals, []string{"evil.snap"})
+	// The file is removed
+	_, err = os.Stat(name)
+	c.Assert(os.IsNotExist(err), Equals, true)
+}
+
+func (s *EnsureDirStateSuite) TestIgnoresUnrelatedFiles(c *C) {
+	name := filepath.Join(s.dir, "unrelated")
+	err := ioutil.WriteFile(name, []byte(`text`), 0600)
+	c.Assert(err, IsNil)
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{})
+	c.Assert(err, IsNil)
+	// Report says that nothing has changed
+	c.Assert(changed, HasLen, 0)
+	c.Assert(removed, HasLen, 0)
+	// The file is still there
+	_, err = os.Stat(name)
+	c.Assert(err, IsNil)
+}
+
+func (s *EnsureDirStateSuite) TestCorrectsFilesWithDifferentSize(c *C) {
+	name := filepath.Join(s.dir, "differing.snap")
+	err := ioutil.WriteFile(name, []byte(``), 0600)
+	c.Assert(err, IsNil)
+	changed, removed, err := osutil.EnsureDirState(s.dir, s.glob, map[string]*osutil.FileState{
+		"differing.snap": {Content: []byte(`Hello World`), Mode: 0600},
+	})
+	c.Assert(err, IsNil)
+	// changed file is reported
+	c.Assert(changed, DeepEquals, []string{"differing.snap"})
+	c.Assert(removed, HasLen, 0)
+	// The content is changed
+	content, err := ioutil.ReadFile(name)
+	c.Assert(err, IsNil)
+	c.Assert(content, DeepEquals, []byte(`Hello World`))
+	// The permissions are what we expect
+	stat, err := os.Stat(name)
+	c.Assert(err, IsNil)
+	c.Assert(stat.Mode().Perm(), Equals, os.FileMode(0600))
+}
+
+func (s *EnsureDirStateSuite) TestCorrectsFilesWithSameSize(c *C) {
```

Comment:
These tests are all almost exactly the same. For the future, these are much better spelled out as a slice of test cases which define the parameters for a single loop that runs the tests.

```
var tests = []testCases{{
    existing: ...
    glob:      ...
    content: ...
}
```

We can even use FileState itself for existing.

But please don't worry about this right now. We've spent way too much time on this functionality already. We'll fix that some day we need to touch this again.